### PR TITLE
misc: improve `asset.py` typing

### PR DESCRIPTION
**Pull request by Victorsitou**
## Summary

Improves `asset.py` typing (soon `member.py`, I need this PR first), the only errors I have are this:

1. 
image
My suggestion is to do `os.PathLike[str]` and `os.PathLike[bytes]`.
See [this](https://stackoverflow.com/a/69680089) and [this](https://github.com/python/typeshed/blob/994b69ef8f18e76689daca3947879c3d7f76173e/stdlib/_typeshed/__init__.pyi#L77-L79).

2. 
image
I don't think we can do anything with this, let's just ignore it?

## Checklist

- [ ] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/asset.py
Comment (shiftinv) [inline disnake/asset.py:204]: classmethods like these were changed from `(cls: Type[T]) -> T` to `(cls) -> Self` a while ago, importing `typing_extensions.Self` in a type-checking-only block (as we still don't explicitly require it). Would you mind updating these methods to use that as well?

Comment (Victorsitou) [reply disnake/asset.py:204]: Sure!

Comment (Victorsitou) [reply disnake/asset.py]: fdec647f0b418c42cb3be3641c0a0eaff4ce9662 ~~forgot I pushed it~~

**Comment by shiftinv** [comment]
Could this PR be unlinked from #889? I don't see anything in there that would be blocking this PR, perhaps they can be merged the other way around? I'm assuming the apidocs PR for avatar decorations won't be merged any time soon.

**Comment by Victorsitou** [comment]
> perhaps they can be merged the other way around? I'm assuming the apidocs PR for avatar decorations won't be merged any time soon.

sounds good for me 👍

**Review by shiftinv — APPROVED** [review]
lgtm!

regarding the remaining linter issues: generic parameters are missing in a couple places all over the code, I started fixing those some time ago but haven't made a PR yet; the url property warning is fine and can safely be ignored, I completely agree.

**Pull request merged**
